### DISCUSSION TITLE
State of code

### DISCUSSION THREAD

**Discussion by ommaury** [LuisaGroup/LuisaRender#25] [General]
Hi, I remember compiling and executing your project on windows successfully when it came out, confirming some of the really good performance numbers presented in the paper. It seems like there has been lot of churn in the code since, and after updating the code, I can not compile it anymore. I don't want to get into to fixing all the errors the current code generates, would it be possible to point me to a commit hash that you know compiled fine? Thanks!

**Comment by Mike-Leo-Smith** [comment]
Hi @ommaury,

Thank you for reaching out regarding the compilation issues you've encountered.

I understand the frustration caused by the changes in the codebase. While I'm eager to assist you, pinpointing a solution or a stable commit without specific compilation errors is challenging. Could you kindly provide some details about the errors you're encountering? This will greatly assist me in identifying the root cause and providing targeted guidance. Additionally, it will benefit other users who may encounter similar challenges with the renderer.

Alternatively, if you're seeking a quick solution, you can download a binary release directly from the [GitHub Actions page](https://github.com/LuisaGroup/LuisaRender/actions/runs/8008910888) (the `LuisaRender-windows-2022-cl` artifact at the bottom of the page).

Your cooperation and input are greatly appreciated. Thank you!

**Reply by ommaury** [reply]
Hi Mike, thank you for taking the time to answer my vague question :) I think I found what I was looking for, i.e. using the `master` branch. I was able to compile and run the scenes. Out of curiosity, is there a roadmap documenting what you are currently working towards?

**Reply by Mike-Leo-Smith** [reply]
Thanks for testing and nice to hear that the `master` branch is working for you.

Unfortunately, we are currently too occupied to create documentation or roadmaps, so I would like to briefly share our progress with you.

We are using the renderer as a testbed, where some of my friends are experimenting with research algorithms such as path guiding, differentiable rendering (with AD), ReSTIR, etc.

Additionally, we are enhancing the underlying LuisaCompute library. Several interesting developments, including a standalone clang-based C++ shading language, a new Vulkan backend, and a refactored IR, are currently ongoing.

**Reply by ommaury** [reply]
Hi Mike,

Thanks to your pointer, looking at the CI logs, I was able to pinpoint the issues I was having, and can now compile and execute the `next` branch `HEAD`. I have 2 questions:

1/ There's a Vulkan backend in the LuisaCompute codebase, though not enabled in the LuisaRender compilation. Is that backend functional?

2/ What is the fastest combination in the current codebase wrt backends/integrators (mega/wave) for RTX hardware on Windows?

Thanks,
Olivier

**Reply by Mike-Leo-Smith** [reply]
Hi, @ommaury 

> 1/ There's a Vulkan backend in the LuisaCompute codebase, though not enabled in the LuisaRender compilation. Is that backend functional?

The `vulkan` backend is still a work in progress and is not functional for LuisaRender.

> 2/ What is the fastest combination in the current codebase wrt backends/integrators (mega/wave) for RTX hardware on Windows?

For RTX-40 series, `cuda` + `mega` should be the fastest for scenes of moderate complexity, while `cuda` + `wave` might be faster if the scene is quite complex or when spectrally rendered. On older platforms (like RTX-20 series), I suppose `dx` + `wave` is still the fastest and `cuda` + `wave` should be only a little bit slower.
